:art: Refactor client and server functions

solareon opened the pull request:
## Description

GetPlayers() is no longer an export in the core. The existing function called overlaps with the native. I renamed the function to FindPlayers() and changed the export to the native.

## Checklist

<!-- Put an x inside the [ ] to check an item, like so: [x] -->

- [X] I have personally loaded this code into an updated Qbox project and checked all of its functionality.
- [X] My code fits the style guidelines.
- [X] My PR fits the contribution guidelines.

solareon committed:
Fix hire menu

Manason commented on server/main.lua:
```diff
@@ -185,11 +185,11 @@ end
 	local playerPed = GetPlayerPed(src)
 	local pCoords = GetEntityCoords(playerPed)
-	for _, v in pairs(exports.qbx_core:GetPlayers()) do
+	for _, v in pairs(GetPlayers()) do
```
Use https://qbox-project.github.io/resources/core/functions/client#getplayersfromcoords

solareon replied:
Got it. Will refactor it today.

solareon committed:
Rework hire menu and fix offline firing

solareon committed:
Fix linter error

solareon commented:
Reworked the hire menu logic to use GetPlayersFromCoords. Took me a minute to figure out you need to use GetPlayerServerId() to return the right id. Also fixed some issues with the offline firing and renamed variables to proper case for locals

Manason commented on client/cl_utils.lua:
```diff
@@ -7,3 +7,14 @@ function comma_value(amount)
+	local playerCoords = GetEntityCoords(cache.ped)
+    local closePlayers = {}
+    for _, v in pairs(GetPlayersFromCoords(playerCoords, 10)) do
+        closePlayers[#closePlayers+1] = GetPlayerServerId(v)
```
v should already be the serverId, is it not?

solareon replied:
No when I was testing I would get 127,128,129 etc. when my server id was 1,2,3 etc.

Manason replied:
Disregard. I see you tested it.

Manason commented on client/cl_utils.lua:
```diff
@@ -7,3 +7,14 @@ function comma_value(amount)
+function FindPlayers()
+	local playerCoords = GetEntityCoords(cache.ped)
+    local closePlayers = {}
+    for _, v in pairs(GetPlayersFromCoords(playerCoords, 10)) do
```
On second thought, if you use https://overextended.dev/ox_lib/Modules/GetNearbyPlayers/Client

Then you can filter out the calling player, since you don't want it in the first place.

solareon replied:
I'll give this a go and see if it eliminates the above call to GetPlayerServerId

Manason replied:
You'll still need to use GetPlayerServerId, but you can get rid of your filtering if condition to filter out the player.

solareon replied:
This did the trick. Still needed the conversion to Server Ids but I got a tighter loop for it.

Manason commented on client/cl_utils.lua:
```diff
@@ -7,3 +7,14 @@ function comma_value(amount)
+    for _, v in pairs(GetPlayersFromCoords(playerCoords, 10)) do
+        closePlayers[#closePlayers+1] = GetPlayerServerId(v)
+    end
+	return lib.callback.await('qb-bossmenu:getplayers', false, closePlayers)
```
```suggestion
	return lib.callback.await('qbx_management:server:getPlayers', false, closePlayers)
```
Would need to be changed elsewhere

solareon replied:
Fixed

Manason commented on server/main.lua:
```diff
@@ -183,35 +183,25 @@ function UpdateGrade(src, data, type)
-			}
-		end
+	for _, v in pairs(closePlayers) do
+		local ped = exports.qbx_core:GetPlayer(v)
```
```suggestion
		local player = exports.qbx_core:GetPlayer(v)
```
A ped is a human/animal GTA object. What you are getting here is the qbx_core player object. They aren't the same thing.

solareon replied:
Fixed

Manason commented on server/main.lua:
```diff
@@ -183,35 +183,25 @@ function UpdateGrade(src, data, type)
+			id = v,
+			name = ped.PlayerData.charinfo.firstname .. " " .. ped.PlayerData.charinfo.lastname,
+			citizenid = ped.PlayerData.citizenid,
+			sourceplayer = ped.PlayerData.source
```
```suggestion
			source = ped.PlayerData.source
```

solareon replied:
Fixed

Manason commented on server/sv_boss.lua:
Get the label the same way you got the payment; don't hardcode.

solareon replied:
Fixed

solareon committed:
Rewrite menus and logic

solareon committed:
Parameter documentation

solareon committed:
Fix linter errors and PlayerData cleanup

solareon commented:
Refactored basically the whole thing. All functions and menus were deduplicated and merged into cl_main. Added parameter documentation for functions (I might have gotten this wrong so please let me know). Fixed typos and variable names.

solareon committed:
Account withdraw bugfix

FjamZoo commented on client/cl_main.lua:
```diff
@@ -0,0 +1,322 @@
+end
+
+---@return table
+function FindPlayers()
+	local playerCoords = GetEntityCoords(cache.ped)
+    local closePlayers = lib.getNearbyPlayers(playerCoords, 10, false)
+    for _, v in pairs(closePlayers) do
+        v.id = GetPlayerServerId(v.id)
+    end
+	return lib.callback.await('qbx_management:server:getplayers', false, closePlayers)
+end
```
```suggestion
local function findPlayers()
    local closePlayers = lib.getNearbyPlayers(GetEntityCoords(cache.ped), 10)
    for _, v in pairs(closePlayers) do
        v.id = GetPlayerServerId(v.id)
    end
    return lib.callback.await('qbx_management:server:getplayers', false, closePlayers)
end
```

I would refactor this function slightly, as it's only used in this file use local variable, if something is not used more than a single time like playerCoords was just put it inside whereever it's needed.

solareon replied:
Did this one mostly for readability but it's still pretty clear what is going on. Changed

FjamZoo commented on client/cl_main.lua:
```diff
@@ -0,0 +1,322 @@
+---@param type 'job'|'gang'
+RegisterNetEvent('qbx_management:client:HireMenu', function(type)
+    local HireMenu = {}
+    local players = FindPlayers()
```
```suggestion
    local players = findPlayers()
```

This would have to be renamed if doing my suggestion above.

solareon replied:
Got it. Changed.

FjamZoo commented on client/cl_main.lua:
```diff
@@ -0,0 +1,322 @@
+    lib.showContext('qbx_management_open_Member')
+end)
+
+---@param type 'job'|'gang'
+RegisterNetEvent('qbx_management:client:HireMenu', function(type)
```
Here I would rename type to, "group" or something that makes it easy for people to see what "type" is as that's very vague, I know the param is there and that's good but we need the variables to have a proper name scheme.

solareon replied:
Changed. Do agree it makes more sense for group vice type.

FjamZoo commented on server/main.lua:
```suggestion
local function AddMoney(account, amount)
```

Since we are going to 1 resource file we need all of these functions to be localized

solareon replied:
Changed. Still getting the hang of Lua best practices.

FjamZoo commented on server/main.lua:
```suggestion
local function RemoveMoney(account, amount)
```

Since we are going to 1 resource file we need all of these functions to be localized

solareon replied:
Changed

FjamZoo commented on server/main.lua:
```suggestion
local function WithdrawMoney(src, amount, type, reason)
```

Since we are going to 1 resource file we need all of these functions to be localized

FjamZoo replied:
I forgot to add the suggestion here but if you make global to local functions make sure to camalCase them

solareon replied:
Changed.

FjamZoo commented on server/main.lua:
like on the client we should rename type here to group, that way its more specific at a glance and we will have similar naming scheme to client and server.

FjamZoo replied:
I forgot to add the suggestion here but if you make global to local functions make sure to camalCase them

solareon replied:
Done and dusted.

FjamZoo commented on server/main.lua:
```suggestion
local function DepositMoney(src, amount, type)
```

Since we are going to 1 resource file we need all of these functions to be localized

FjamZoo replied:
I forgot to add the suggestion here but if you make global to local functions make sure to camalCase them

solareon replied:
Also done. Fixed as few others where the camelcase rule wasn't applied

solareon replied:
Is there a preference for net events to be lowercase or camelCase?

BerkieBb replied:
camelCase

FjamZoo commented on server/main.lua:
```suggestion
lib.callback.register('qbx_management:server:getemployees', function(source, accountName, type)
	return GetEmployees(source, accountName, type)
end)
```

When doing callbacks on the server we will always have the source as an arg, which means we do not need to localize it.

solareon replied:
What about the cases where we don't use source. is it okay to discard it with _ ?

Manason replied:
Yes. It is preferred to discard it with "_" when not using it.

FjamZoo commented on server/main.lua:
```suggestion
RegisterNetEvent('qbx_management:server:GradeUpdate', function(data)
	UpdateGrade(source, data)
	TriggerClientEvent('qbx_management:client:OpenMenu', source, data.type)
end)
```

We only want and need to localize source if the coroutine was yielded.

Which in layman terms mean if we use a Wait anywhere

solareon replied:
done.

FjamZoo commented on server/main.lua:
```suggestion
---@param group 'job'|'gang'
---@param amount number
RegisterNetEvent('qbx_management:server:withdrawMoney', function(group, amount)
	local success, player, account = WithdrawMoney(source, amount, group, group == 'gang' and 'Gang menu withdraw' or 'Boss menu withdraw')
	if not success or not player then return end
	local logArea = group == 'gang' and 'gang' or 'boss'
	
	TriggerEvent('qb-log:server:CreateLog', logArea..'menu', 'Withdraw Money', 'yellow', player.PlayerData.charinfo.firstname .. ' ' .. player.PlayerData.charinfo.lastname .. ' successfully withdrew $' .. amount .. ' (' .. account .. ')', false)
	TriggerClientEvent('qbx_management:client:OpenMenu', source, type)
end)
```

here we want to rename type to group, to keep going with the naming scheme above.

I also removed the local src = source here because we do not have a yield in any of these functions.

solareon replied:
done

FjamZoo commented on server/main.lua:
```suggestion
---@param group 'job'|'gang'
---@param amount number
RegisterNetEvent('qbx_management:server:depositMoney', function(group, amount)
	local success, player, account = DepositMoney(source, amount, group)
	if not success or not player then return end

	local logArea = group == 'gang' and 'gang' or 'boss'
	TriggerEvent('qb-log:server:CreateLog', logArea..'menu', 'Deposit Money', 'yellow', player.PlayerData.charinfo.firstname .. ' ' .. player.PlayerData.charinfo.lastname .. ' successfully deposited $' .. amount .. ' (' .. account .. ')', false)
	TriggerClientEvent('qbx_management:client:OpenMenu', source, group)
end)
```

here we want to rename type to group, to keep going with the naming scheme above.

I also removed the local src = source here because we do not have a yield in any of these functions.

solareon replied:
Done

FjamZoo commented on client/cl_main.lua:
```diff
@@ -0,0 +1,322 @@
+	return lib.callback.await('qbx_management:server:getplayers', false, closePlayers)
+end
+
+---@param type 'job'|'gang'
+RegisterNetEvent('qbx_management:client:SocietyMenu', function(type)
+    local amount = lib.callback.await('qbx_management:server:getAccount', false, QBX.PlayerData[type].name)
+    local SocietyMenu = {
+        {
+            title = 'Deposit',
+            icon = 'fa-solid fa-money-bill-transfer',
+            description = 'Deposit Money',
+            event = 'qbx_management:client:SocietyDeposit',
```
Here and downwords there's a bunch of "type" that we need to have renamed to group as you have seen in the other messages.

solareon replied:
Done.

FjamZoo reviewed (changes requested):
There's a lot of changes requested here, I could of pointed out everywhere that has "type" that we need to rename to something more appropriate like "group" but I think you get the idea from all the suggestions I've made.

Manason commented:
I'll add that in the future, please keep your PRs small. This PR is difficult to review now because of it's size.

solareon committed:
Localize functions and clean up

solareon commented:
> I'll add that in the future, please keep your PRs small. This PR is difficult to review now because of it's size.

My apologies for the large PR. Would it help if I squashed the commits down? I didn't intend to rewrite the whole thing but kind of got in the groove after the first couple changes.

I made all the changes regarding the camelCase, localizing functions where appropriate, changing type to group, and only localizing source where appropriate.

solareon committed:
Linter error fix

Manason commented on server/main.lua:
```suggestion
lib.callback.register('qbx_management:server:getEmployees', function(source, accountName, group)
```

Manason commented on server/main.lua:
Should this be a callback returning nil instead? Events should describe something that has happened in the past https://manason.github.io/effective-fivem-lua/events/#past-tense

Manason commented on server/main.lua:
```suggestion
RegisterNetEvent('qbx_management:server:hireEmployee', function(data)
```

Manason commented on server/main.lua:
Ditto on callback

Manason commented on server/main.lua:
```suggestion
lib.callback.register('qbx_management:server:getPlayers', function(_, closePlayers)
```

Manason commented on server/main.lua:
Optional; Consider using string.format to make this a bit easier to read
```suggestion
        exports.qbx_core:Notify(source, 'You hired %s into %s':format(targetFullName, organizationLabel), 'success')
```

Manason commented on server/main.lua:
Was this previously not localized?

solareon replied:
No. There was no locales in this resource.

solareon replied:
No there wasn't any translations in this resource.

Manason commented on server/main.lua:
This is a large function. Consider if it can be broken up into smaller functions.

Manason commented on server/main.lua:
Seems like this should be a callback

Manason commented on server/main.lua:
Also seems like it should be a callback

Manason commented on client/cl_main.lua:
```diff
@@ -0,0 +1,316 @@
+local jobs = exports.qbx_core:GetJobs()
+local gangs = exports.qbx_core:GetGangs()
+
+local function commaValue(amount)
```
I think core utils has a comma value util?

Manason commented on client/cl_main.lua:
```diff
@@ -0,0 +1,316 @@
+    local hireName = QBX.PlayerData[group].name
+    for _, v in pairs(players) do
+
+        if v and v.citizenid ~= QBX.PlayerData.citizenid and v[group].name ~= hireName then
```
players will no longer include the client, so no need to check that citizen is not equals. Additionally, it is impossible for the value of a table loops through using pairs to be nil.
```suggestion
        if v[group].name ~= hireName then
```

Manason commented on client/cl_main.lua:
```diff
@@ -0,0 +1,316 @@
+    local hireMenu = {}
+    local players = findPlayers()
+    local hireName = QBX.PlayerData[group].name
+    for _, v in pairs(players) do
```
Consider a more descriptive variable name than 'v'
```suggestion
    for _, player in pairs(players) do
```

Manason commented:
> > I'll add that in the future, please keep your PRs small. This PR is difficult to review now because of it's size.
> 
> My apologies for the large PR. Would it help if I squashed the commits down? I didn't intend to rewrite the whole thing but kind of got in the groove after the first couple changes.
> 
> I made all the changes regarding the camelCase, localizing functions where appropriate, changing type to group, and only localizing source where appropriate.

That would make it worse actually. In the future please keep your PRs focused on one change. You can branch off your branch to create a second PR, and branch off of that branch to create a third, etc..

solareon commented:
Closing this PR to submit a new one due to full rewrite.
